### ISSUE TITLE
Feature Request: Automatically assign random color to new instances

### ISSUE THREAD

**Issue by deadlydog** [johnpapa/vscode-peacock#33]
Hey John, first off, great extension :)

I know that you can already assign a default color per repo that gets stored in the `.vscode/settings.json` file, but that's not ideal for a few reasons:
- I have 30+ repos and don't want to have to try and assign them each a unique color.
- I typically prefer my theme's default color, and really only want a different color when I have more than one instance open.
- For shared repos, many people don't like committing the `.vscode/settings.json` file as it often contains personal settings that not all users agree upon.

Because of this, it would be great if the extension was able to automatically detect if it's the only instance open or not. If it's the only one, then it doesn't do anything, but for each subsequent instance that's opened it would auto-assign a different color to the window when it's opened. Picking a random color for each subsequent instance would probably be easiest, but bonus points if there was a setting to specify a list of preferred colors that it picked from. Even more bonus points if you were able to actually specify 2nd instance is always color X, 3rd window is always color Y, 4th window is always color Z, etc.

**Comment by johnpapa** [comment]
interesting idea. how would you want it to work? set a setting for peacock such as `peacock.chooseRandomColorOnOpen = true` and then when you open a project it checks if you have a color (if so it does nothing) and if you dont, it adds one?


hmmmmmm

**Comment by deadcoder0904** [comment]
I was thinking about this feature & was gonna make the same extension with this same feature when I saw [Wes Bos doing it](https://twitter.com/addyosmani/status/1091962556095188992?lang=en) but you beat me to it 😉

What I wanted was 2 things -

1. Choose Random Color when new VSCode instance gets opened

2. Have a Set Of Predefined Colors because not all colors look good on title bars, etc..

Would love to see this implemented :)

**Comment by deadlydog** [comment]
> interesting idea. how would you want it to work? set a setting for peacock such as `peacock.chooseRandomColorOnOpen = true` and then when you open a project it checks if you have a color (if so it does nothing) and if you dont, it adds one?

Yup, that's pretty much it. And @deadcoder0904 hit on the 2 primary features I mentioned that would be awesome.

The only additional feature that I mentioned that I would _really_ like if possible is:

- Only apply a color if there's more than one instance open.

I'm not sure if you're able to detect if there's more than one instance open or not though, especially since VS Code is cross platform. It wouldn't be the worst thing in the world if it always applied a color scheme, but 80% of the time I'm only working in a single instance and wouldn't need the different color scheme.

**Comment by johnpapa** [comment]
neat idea. i dont think i see this as part of the roadmap for now. thanks tho

**Comment by sheldonhull** [comment]
I'd like to chime in that my google feed showed a post on this topic and I thought I'd give this a try. I really like the customization/concept, however the ability to automatically assign is important to me for new workspaces as it's another step I don't have to worry about. 

I'd mention the other extension I'm using that was before I saw your post and does this is: `stuart.unique-window-colors` . This allows automatically setting, but you can still override. Perhaps consider something similar as it's very much a similar extension. Thanks for sharing this, look forward to more improvements!

**Comment by johnpapa** [comment]
To consider this I'd like to see the plans for how this would work and be tested.

**Comment by deadlydog** [comment]
Awesome, thanks for sharing that @sheldonhull . Your comment helped me find the `Window Colors` extension you were referring to. That is almost exactly what I'm looking for, except that extension _always_ modifies the `.vscode/settings.json` file which is undesirable. It means as soon as I open a workspace it wants to modify that setting, so I'll need to explicitly add it to all of my .gitignore files in all of my various repos, which is a bit annoying.

I find this extension's implementation to be nicer since it doesn't _have_ to touch that file; it's just missing the one feature that would make make it truly great; eliminating the need for me to manually apply a color every time I open a new workspace.

**Comment by deadcoder0904** [comment]
> Have a Set Of Predefined Colors because not all colors look good on title bars, etc..

Thanks for implementing 👏

**Comment by johnpapa** [comment]
re-opening for consideration

**Comment by Nivmorabin** [comment]
Hi,

Sorry for the duplicate #81  , didn't seem to find this feature request.

It would be great to have this option.

Thanks in advance.

**Comment by johnpapa** [comment]
No worries. I had closed it but recently repopened. Thanks for reporting this

**Comment by musicfuel** [comment]
@johnpapa I've got this working from a setting and functional perspective as part of the activate call for the extension, but I can't figure out a good way to test it. When running tests in the Extension Host, the activate seems to occur before any of the tests are run and it doesn't appear that manually calling activate in a test invokes the call where the logic resides (likely because it is being bypassed because VS Code knows the extension is already active). I've tried a few different approaches, but I can't figure out the right derivation of suite setup calls to prepare a configuration and then have activate logic run.

Any thoughts? Am I approaching this problem in a wrong or naive way where I should not be applying the random color as part of activate?

**Comment by johnpapa** [comment]
Thinking out loud ...

The logic, if isolated, could be executed in a test faking the activate. It checks for a config setting `surpriseMeOnActivate` and if `true` and there isn't already a color set, it sets one. The test could just  test the feature.

Is it necessary to test this on activate? Possibly. Could a test check to see if after activate there are colors in the file? It should be empty (setup?) and if it has the settings, you could assert after startup.

What do you think?

**Comment by musicfuel** [comment]
That all sounds like a plan. We probably don't have to test that it occurs on activate as much as it occurs in response to the setting. I will work with that. Thanks for the guidance!

**Comment by johnpapa** [comment]
Sounds great. 

Note that I updated the code to use webpack bundling today. So be aware the testing works but I’m not 100% sure I tuned it right. Just in case it gives you trouble. It’s in master along with some new logging. I marked it for 1.2.2 but I’d we add this feature I’d move to 1.3 according to semver

**Comment by deadlydog** [comment]
> neat idea. i dont think i see this as part of the roadmap for now. thanks tho

Thank you for reconsidering and incorporating this feature request John 👍

**Comment by johnpapa** [comment]
it's not everything you asked for, but i think this is the best we can do right now without more API support.

thanks for the ideas!

**Comment by Samir-romdhani** [comment]
with `"peacock.surpriseMeOnStartup": true` it works

**Comment by iyerusad** [comment]
@johnpapa Now this is in, two small suggestions regarding discoverability:

1. **Rename `surpriseMeOnStartup` to `AutomaticMode` or something similar.** My initial run through peacock settings saw this flag but the description sounded kinda crazy "Specifies that Peacock should surprise you at the start of every editing session"... With what?? Auto injecting semicolons? LOL. Tldr; "AutomaticMode" seems more understandable than "SurpriseMe" for this feature.

1. **Enable `surpriseMeOnStartup`/`AutomaticMode` by default/new users**. I started regularly launching multiple repos (right click open with VSCode) and went looking for a window color shifter. I installed/uninstalled peacock 2-3 times because it "wasn't working", maybe its because i didn't have `.codeworkspace` file... No - by default surprise/automatic mode is off. Would throw hat in that it should be on by default for new users like myself who little slow on the uptake.

That aside, thank you for adding this feature. Makes the extension a must have in dotfiles default for me.
